File: backend/app/api/tips.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.models.models import User as UserModel, Expense as ExpenseModel, TipMethod
from app.schemas.schemas import TipCreate
from app.services.utils import get_current_week_id
from app.api.deps import check_role
from app.models.models import UserRole
from datetime import datetime

router = APIRouter()

@router.post("/")
def create_tip(
    tip_in: TipCreate,
    db: Session = Depends(get_db),
    current_user = Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))
):
    employee = db.query(UserModel).filter(UserModel.id == tip_in.employee_id).first()
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")

    if tip_in.method == TipMethod.WAGES:
        employee.payable_balance += tip_in.amount
    else:
        week_id = get_current_week_id()
        expense = ExpenseModel(
            description=f"Chai - {employee.full_name}",
            amount=tip_in.amount,
            category="Chai",
            week_id=week_id
        )
        db.add(expense)

    db.commit()
    return {"message": "Tip logged successfully", "method": tip_in.method.value, "employee": employee.full_name}

File: server/src/routes/tips.ts
import { Router } from 'express';
import { pool } from '../db';
import { HttpError, getCurrentWeekId, wrap } from '../utils';
import { asFloat, asOptStr } from '../finance';
import { requireRole } from '../auth';
import { ROLES } from '../types';

const router = Router();
const M = ROLES.slice(0, 2); // ADMIN, MANAGER

// POST /tips
router.post(
  '/',
  wrap(async (req, res) => {
    await requireRole(req, M);
    const body = req.body || {};
    const employee_id = Number(body.employee_id);
    if (!Number.isInteger(employee_id)) throw new HttpError(422, 'employee_id must be an integer');
    const amount = asFloat(body.amount, 'amount');
    if (amount < 0) throw new HttpError(422, 'amount must be >= 0');
    const method = (asOptStr(body.method) ?? '').toLowerCase();
    if (method !== 'cash' && method !== 'wages') {
      throw new HttpError(422, 'value is not a valid enumeration member for method');
    }

    const emp = (
      await pool.query('SELECT id, full_name, payable_balance FROM users WHERE id = $1', [employee_id])
    ).rows[0];
    if (!emp) throw new HttpError(404, 'Employee not found');
    if (!emp.full_name) throw new HttpError(404, 'Employee not found');

    if (method === 'wages') {
      // WAGES = add to wage balance
      await pool.query('UPDATE users SET payable_balance = payable_balance + $1 WHERE id = $2', [
        amount,
        employee_id,
      ]);
    } else {
      // CASH = record a Chai expense
      const week_id = getCurrentWeekId();
      const now = new Date().toISOString().replace('T', ' ').replace('Z', '');
      await pool.query(
        `INSERT INTO expenses (timestamp, description, amount, category, week_id)
         VALUES ($1,$2,$3,$4,$5)`,
        [now, `Chai - ${emp.full_name}`, amount, 'Chai', week_id],
      );
    }

    res.json({ message: 'Tip logged successfully', method, employee: emp.full_name });
  }),
);

export default router;